features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/CaffeineDnsCache.java

```diff
@@ -0,0 +1,370 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package org.opennms.netmgt.dnsresolver.netty;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static io.netty.util.internal.ObjectUtil.checkPositiveOrZero;
+
+import java.net.InetAddress;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.base.MoreObjects;
+
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.dns.DnsPtrRecord;
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.resolver.dns.DnsCache;
+import io.netty.resolver.dns.DnsCacheEntry;
+import io.netty.util.internal.StringUtil;
+
+/**
+ * DNS cache implementation largely copied from Netty's
+ *   https://github.com/netty/netty/blob/netty-4.1.38.Final/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java
+ * but adapted slightly to work with our {@link CaffeineCache}.
+ */
+public class CaffeineDnsCache implements ExtendedDnsCache {
+
+    protected static final int MAX_SUPPORTED_TTL_SECS = (int) TimeUnit.DAYS.toSeconds(365 * 2);
+    protected static final int DEFAULT_NEGATIVE_TTL_SECS = (int) TimeUnit.MINUTES.toSeconds(5);
+    protected static final int DEFAULT_MAX_SIZE = 10000;
+
+    private final CaffeineCache<ExtendedDnsCacheEntry> resolveCache;
+
+    private final int minTtl;
+    private final int maxTtl;
+    private final int negativeTtl;
+    private final long maxSize;
+
+    // Track hits and misses ourselves since the underlying cache doesn't expose these
+    private final Meter cacheHits = new Meter();
+    private final Meter cacheMisses = new Meter();
+
+    /**
+     * Create a cache that respects the TTL returned by the DNS server
+     * and doesn't cache negative responses.
+     */
+    public CaffeineDnsCache() {
+        this(0, MAX_SUPPORTED_TTL_SECS, DEFAULT_NEGATIVE_TTL_SECS, DEFAULT_MAX_SIZE);
+    }
+
+    /**
+     * Create a cache.
+     * @param minTtl the minimum TTL
+     * @param maxTtl the maximum TTL
+     * @param negativeTtl the TTL for failed queries
+     */
+    public CaffeineDnsCache(int minTtl, int maxTtl, int negativeTtl, long maxSize) {
+        this.minTtl = Math.min(MAX_SUPPORTED_TTL_SECS, checkPositiveOrZero(minTtl, "minTtl"));
+        this.maxTtl = Math.min(MAX_SUPPORTED_TTL_SECS, checkPositiveOrZero(maxTtl, "maxTtl"));
+        if (minTtl > maxTtl) {
```

Comment:
This can be simplified by using guavas `Precondition.checkArgument`.

Reply:
Done.